[FLINK-10942][network,test] Deduplicate common codes in OutputEmitterTest

zhijiangW opened the pull request:
## What is the purpose of the change

*There are many duplicated codes in `OutputEmitterTest` to make it difficult to maintain. So it is necessary to abstract the common codes to make it simple which brings benefits for the following refactor work in `ChannelSelector` interface.*

## Brief change log

  - *Abstract common codes for `OutputEmitterTest`*

## Verifying this change

This change is already covered by existing tests, such as *OutputEmitterTest*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/operators/util/OutputEmitterTest.java:
```diff
@@ -48,404 +49,272 @@
+			ShipStrategyType.PARTITION_HASH, testIntComp);
+		final SerializationDelegate<Integer> serializationDelegate = new SerializationDelegate<>(new IntSerializer());
+
 		// MinVal hash
```
nit: I would drop `// MinVal hash`, `// -1 hash`, `// ... hash` comments here. They don't explain anything more then the `assertPartitionHashSelectedChannels` method name already does.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/operators/util/OutputEmitterTest.java:
```diff
@@ -48,404 +49,272 @@
+	private void verifyPartitionHashSelectedChannels(int numRecords, int numChannels, Either<Integer, String> recordType) {
+		int[] hits = getSelectedChannels(ShipStrategyType.PARTITION_HASH, numRecords, numChannels, recordType);
+
+		int cnt = 0;
```
nit: rename to `totalHitCount`?

zhijiangW replied:
Agree, the current abbreviate is less semantics.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/operators/util/OutputEmitterTest.java:
```diff
@@ -48,404 +49,272 @@
+		ShipStrategyType shipStrategyType,
+		int numRecords,
+		int numChannels,
+		Either<Integer, String> recordType) {
```
formatting nit: could you add one more tab before the parameters? As it is now, parameters are fusing with the method code.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/operators/util/OutputEmitterTest.java:
```diff
@@ -48,404 +49,272 @@
+		return getSelectedChannels(selector, delegate, recordType, numRecords, numChannels);
+	}
+
+	private int[] getSelectedChannels(
```
nit: can you reorder the `getSelectedChannels` methods so that all of them are besides one another, and the "most private" one is the last? I mean if `foo(int)` is calling `foo(int, long)`, `foo(int)` should be before `foo(int, long)`? I think this is more or less style that we are following and it helps to avoid jumping up & down when looking for the implementation.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/operators/util/OutputEmitterTest.java:
```diff
@@ -48,404 +49,272 @@
+	private int[] getSelectedChannels(
+		ChannelSelector<SerializationDelegate<Record>> selector,
+		SerializationDelegate<Record> delegate,
+		Either<Integer, String> recordType,
```
I think you are never using the value of `recordType` anywhere? If that's a place holder only for selecting the type, I would switch it to simple `enum RecordType { INTEGER, STRING }`. As it is now, it's a bit confusing for example **"why 0?"** when you call:
```
getSelectedChannels(selector, delegate, new Either.Left<>(0), numRecords, numChannels)
```

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/operators/util/OutputEmitterTest.java:
```diff
@@ -48,404 +49,272 @@
+		int[] hits = getSelectedChannels(ShipStrategyType.FORWARD, numRecords, numChannels, recordType);
+
+		assertTrue(hits[0] == numRecords);
+		for (int i = 1; i < hits.length; i++) {
```
nit: `for (int hit : hits)`?

zhijiangW replied:
It is from index 1 not 0, so not make sense for short loop way.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/operators/util/OutputEmitterTest.java:
```diff
@@ -48,404 +49,272 @@
+	}
+
+	private void verifyPartitionHashSelectedChannels(int numRecords, int numChannels, Either<Integer, String> recordType) {
+		int[] hits = getSelectedChannels(ShipStrategyType.PARTITION_HASH, numRecords, numChannels, recordType);
```
nit: maybe rename `getSelectedChannels` to `getSelectedChannelsHitCount`? It's indeed difficult to name this method, but `int[] hits = getSelectedChannels()` looks strange and I had to go through the implementation to understand what is happening :)

zhijiangW commented:
@pnowojski , thanks for helpful reviews!
I almost agree with all your above comments except the short loop way I mentioned inline.
I already submit another hotfix for addressing all the issues, wish not missing anything. :)
